File: overlay-prototype/packages/grant/canonical.ts
/**
 * The restricted JCS subset from techspecsigner.md §4.
 *
 * The Swift port in signer/Sources/GrantSigner/Canonical.swift must reproduce
 * this byte for byte. CanonicalTests.swift proves it against vectors/canonical.json.
 */

import type { JsonValue } from "./types.ts";

export type CanonicalErrorReason =
  | "null"
  | "float"
  | "nan"
  | "infinity"
  | "unsafe_integer"
  | "lone_surrogate"
  | "unsupported";

export class CanonicalError extends Error {
  readonly reason: CanonicalErrorReason;
  readonly path: string;

  constructor(reason: CanonicalErrorReason, path: string, detail: string) {
    super(`canonicalize: ${detail} at ${path || "<root>"}`);
    this.name = "CanonicalError";
    this.reason = reason;
    this.path = path;
  }
}

const TWO_CHAR_ESCAPES: Record<string, string> = {
  '"': '\\"',
  "\\": "\\\\",
  "\b": "\\b",
  "\f": "\\f",
  "\n": "\\n",
  "\r": "\\r",
  "\t": "\\t",
};

/**
 * §4: the listed characters become two-character sequences, other code points
 * below 0x20 become \u00xx with LOWERCASE hex, everything else stays raw.
 */
function encodeString(value: string, path: string): string {
  let out = '"';
  for (let i = 0; i < value.length; i += 1) {
    const char = value[i];
    const code = value.charCodeAt(i);
    const escape = TWO_CHAR_ESCAPES[char];
    if (escape !== undefined) {
      out += escape;
    } else if (code < 0x20) {
      out += `\\u${code.toString(16).padStart(4, "0")}`;
    } else if (code >= 0xd800 && code <= 0xdfff) {
      // A lone surrogate has no UTF-8 encoding, so JavaScript and Swift would
      // silently disagree on the bytes. §4's rule is to throw, never coerce.
      const isHigh = code <= 0xdbff;
      const next = isHigh ? value.charCodeAt(i + 1) : Number.NaN;
      if (isHigh && next >= 0xdc00 && next <= 0xdfff) {
        out += char + value[i + 1];
        i += 1;
      } else {
        throw new CanonicalError(
          "lone_surrogate",
          path,
          `unpaired surrogate \\u${code.toString(16).padStart(4, "0")}`,
        );
      }
    } else {
      out += char;
    }
  }
  return `${out}"`;
}

/** §4: integers only — no exponent, no decimal point, no sign on positives. */
function encodeNumber(value: number, path: string): string {
  if (Number.isNaN(value)) throw new CanonicalError("nan", path, "NaN is not encodable");
  if (!Number.isFinite(value)) {
    throw new CanonicalError("infinity", path, "Infinity is not encodable");
  }
  if (!Number.isInteger(value)) {
    throw new CanonicalError("float", path, `${value} is not an integer`);
  }
  if (!Number.isSafeInteger(value)) {
    // Beyond 2^53 JavaScript cannot round-trip the value, so the Swift Int64
    // port would print different digits.
    throw new CanonicalError("unsafe_integer", path, `${value} exceeds the safe integer range`);
  }
  // String(-0) is "0", which is what we want.
  return String(value);
}

function encode(value: unknown, path: string): string {
  if (value === null) {
    throw new CanonicalError("null", path, "null is not encodable");
  }
  switch (typeof value) {
    case "string":
      return encodeString(value, path);
    case "number":
      return encodeNumber(value, path);
    case "boolean":
      return value ? "true" : "false";
    case "object":
      break;
    default:
      throw new CanonicalError("unsupported", path, `${typeof value} is not encodable`);
  }

  if (Array.isArray(value)) {
    // §4: arrays keep their order.
    const items = value.map((item, index) => encode(item, `${path}[${index}]`));
    return `[${items.join(",")}]`;
  }

  if (Object.getPrototypeOf(value) !== Object.prototype && Object.getPrototypeOf(value) !== null) {
    throw new CanonicalError("unsupported", path, "only plain objects are encodable");
  }

  // §4: keys sorted by UTF-16 code unit. Array.prototype.sort() compares
  // strings exactly that way, so the default sort is already correct here.
  // It is the Swift port that needs an explicit Array(s.utf16) comparison.
  const keys = Object.keys(value as Record<string, unknown>).sort();
  const entries = keys.map((key) => {
    const child = (value as Record<string, unknown>)[key];
    const childPath = path ? `${path}.${key}` : key;
    if (child === undefined) {
      throw new CanonicalError("unsupported", childPath, "undefined is not encodable");
    }
    return `${encodeString(key, childPath)}:${encode(child, childPath)}`;
  });
  // §4: no whitespace anywhere.
  return `{${entries.join(",")}}`;
}

/** The canonical string whose UTF-8 bytes are signed and hashed. */
export function canonicalize(value: JsonValue): string {
  return encode(value, "");
}

export function base64url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

export function base64urlDecode(value: string): Uint8Array {
  const padded = value.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(padded.padEnd(Math.ceil(padded.length / 4) * 4, "="));
  return Uint8Array.from(binary, (char) => char.charCodeAt(0));
}

export async function sha256(bytes: Uint8Array): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.digest("SHA-256", bytes));
}

/** base64url(SHA-256(canonicalize(value))) — the form every hash in §3 takes. */
export async function canonicalHash(value: JsonValue): Promise<string> {
  const bytes = new TextEncoder().encode(canonicalize(value));
  return base64url(await sha256(bytes));
}

File: overlay-prototype/signer/Sources/GrantCore/Canonical.swift
import Foundation

/// §4: object keys sort by UTF-16 code unit.
///
/// Swift's default `String` `<` is Unicode *scalar* ordering, which disagrees with
/// JavaScript for anything above the BMP: U+1F600 has scalar value 128512 and
/// sorts after U+FF5A (65370), but its first UTF-16 code unit is 0xD83D (55357)
/// and sorts before. Compare the code units directly.
@inline(__always)
public func utf16Less(_ a: String, _ b: String) -> Bool {
    var i = a.utf16.makeIterator()
    var j = b.utf16.makeIterator()
    while true {
        switch (i.next(), j.next()) {
        case (nil, nil): return false       // equal
        case (nil, _?): return true         // a is a prefix of b
        case (_?, nil): return false
        case let (x?, y?):
            if x != y { return x < y }
        }
    }
}

private let twoCharEscapes: [UInt16: [UInt8]] = [
    0x22: Array("\\\"".utf8),
    0x5C: Array("\\\\".utf8),
    0x08: Array("\\b".utf8),
    0x0C: Array("\\f".utf8),
    0x0A: Array("\\n".utf8),
    0x0D: Array("\\r".utf8),
    0x09: Array("\\t".utf8),
]

private let lowerHex: [UInt8] = Array("0123456789abcdef".utf8)

/// §4: the seven listed escapes, other sub-0x20 code points as `\u00xx` with
/// LOWERCASE hex, every other code point raw.
func encodeString(_ value: String, path: String, into out: inout [UInt8]) throws {
    out.append(0x22)
    let units = Array(value.utf16)
    var index = 0
    while index < units.count {
        let unit = units[index]
        if let escape = twoCharEscapes[unit] {
            out.append(contentsOf: escape)
        } else if unit < 0x20 {
            out.append(contentsOf: Array("\\u00".utf8))
            out.append(lowerHex[Int((unit >> 4) & 0xF)])
            out.append(lowerHex[Int(unit & 0xF)])
        } else if unit >= 0xD800 && unit <= 0xDFFF {
            // A lone surrogate has no UTF-8 encoding; emitting one would make the
            // two languages disagree on bytes. §4's rule is to throw, never coerce.
            let isHigh = unit <= 0xDBFF
            let next: UInt16? = index + 1 < units.count ? units[index + 1] : nil
            guard isHigh, let low = next, low >= 0xDC00, low <= 0xDFFF else {
                throw CanonicalError(
                    .loneSurrogate, path: path,
                    detail: String(format: "unpaired surrogate \\u%04x", Int(unit)))
            }
            var scalarValue = UInt32(unit - 0xD800) << 10
            scalarValue += UInt32(low - 0xDC00)
            scalarValue += 0x1_0000
            out.append(contentsOf: Array(String(UnicodeScalar(scalarValue)!).utf8))
            index += 2
            continue
        } else {
            out.append(contentsOf: Array(String(UnicodeScalar(unit)!).utf8))
        }
        index += 1
    }
    out.append(0x22)
}

private func encode(_ value: CanonicalValue, path: String, depth: Int, into out: inout [UInt8]) throws {
    guard depth <= canonicalMaxDepth else {
        throw CanonicalError(.depthExceeded, path: path, detail: "nesting deeper than \(canonicalMaxDepth)")
    }
    switch value {
    case .null:
        throw CanonicalError(.null, path: path, detail: "null is not encodable")
    case .boolean(let flag):
        out.append(contentsOf: Array((flag ? "true" : "false").utf8))
    case .integer(let number):
        guard abs(number) <= canonicalMaxSafeInteger else {
            throw CanonicalError(.unsafeInteger, path: path,
                                 detail: "\(number) exceeds the safe integer range")
        }
        // §4: no sign on positives, no exponent, no decimal point.
        out.append(contentsOf: Array(String(number).utf8))
    case .string(let text):
        try encodeString(text, path: path, into: &out)
    case .array(let items):
        out.append(0x5B) // [
        for (offset, item) in items.enumerated() {
            if offset > 0 { out.append(0x2C) }
            try encode(item, path: "\(path)[\(offset)]", depth: depth + 1, into: &out)
        }
        out.append(0x5D) // ]
    case .object(let pairs):
        // §4: sorted by UTF-16 code unit, no whitespace anywhere.
        let sorted = pairs.sorted { utf16Less($0.key, $1.key) }
        out.append(0x7B) // {
        for (offset, pair) in sorted.enumerated() {
            if offset > 0 { out.append(0x2C) }
            let childPath = path.isEmpty ? pair.key : "\(path).\(pair.key)"
            try encodeString(pair.key, path: childPath, into: &out)
            out.append(0x3A) // :
            try encode(pair.value, path: childPath, depth: depth + 1, into: &out)
        }
        out.append(0x7D) // }
    }
}

/// The canonical bytes. Built as UTF-8 directly rather than as a `String` that is
/// converted at the end, so no `String` normalization can touch the output.
public func canonicalBytes(_ value: CanonicalValue) throws -> [UInt8] {
    var out: [UInt8] = []
    out.reserveCapacity(256)
    try encode(value, path: "", depth: 0, into: &out)
    return out
}

public func canonicalString(_ value: CanonicalValue) throws -> String {
    String(decoding: try canonicalBytes(value), as: UTF8.self)
}
